search-index/text2embeddings.py

```diff
@@ -18,10 +19,108 @@
+    # TODO: Note, we can potentially do this in a more robust way by using the text block custom attributes created
+    #  in postprocessor.py, which I've updated. This will allow us to deal with edge cases, for example where the
+    #  previous block is not either context or part of the same list from the previous page. This would be the case
+    #  when, for example, there are random elements between pages that the postprocessors haven't been able to catch.
+    #  But I've ignored this for now because it seems that the postprocessing in postprocessor.py has dealt with the
+    #  vast majority of edge cases (as indicated by the rarity of certain debugging metadata).
+    text = prev_processed_text_block + "\n" + text
+    return text
+
+
+def delete_string_indices(data: str, indices: List[int]) -> str:
+    """Delete a list of indices from a string.
+    Args:
+        data (str): string to delete indices from.
+        indices (List[int]): list of indices to delete.
+    Returns:
+        str: string with indices deleted.
+    """
+    return "".join([char for idx, char in enumerate(data) if idx not in indices])
+
+
+def get_text_from_merged_block(text_block: dict) -> str:
+    """
+    Remove unnecessary styling from merged text blocks (superscripts).
+
+    Args:
+        text_block: A block of text that has been merged.
+
+    Returns:
+        str: The text block with styling that is not related to semantics removed.
+
+    """
+    # Text output with no processing.
+    text_output = (
+        "\n".join(text_block["text"]).strip(),
+        text_block["text_block_id"],
+    )
+    # Remove superscripts only. Bolding has no impact on the block's content once merged, so we don't need to remove
+    # anything. Subscripts are kept because, for example, we want to keep CO2 even when it is subscripted).
+    # Superscripts are almost always references not related to semantics, so we can remove them.
+    try:
+        style_spans = text_block["custom_attributes"]["styleSpans"]
+        delete_indices = []
+        for style_span in style_spans:
+            if style_span["style"] == "superscript":
+                remove_start = style_span["start_idx"]
+                remove_end = style_span["end_idx"]
+                # if superscript is of the form st, nd, rd, th, do not remove as this can potentially
+                # be used by language models.
+                delete_substr = text_output[0][remove_start : remove_end + 1]
+                if re.match(r"(st|nd|rd|th)$", delete_substr):
+                    continue  # don't remove
+                delete_indices.extend([remove_start, remove_end])
+
+        # The indices are relative to the whole block, not just elements in the block.
+        # Remove the indices from deleted_indices from text_output.
+        if delete_indices:  # There will only be indices to delete in superscript case.
+            text_output_amended = delete_string_indices(text_output[0], delete_indices)
```

Comment:
It seems like you're passing start, end pairs into this function but the function actually deletes characters by index? Should the `delete_indices.extend` line add a range of indices from start to end instead?

Reply:
Good spot. Since I was only really dealing with superscripts and mostly cases of a single superscript, hadn't spotted it!